HPCC-18132 - ESDL support include paths for searching EspInclude files

mayx opened the pull request:
- Add the -i --includePath options to relevant esdl commands
- Add support in esdl compiler to search included files in the include
  paths

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

mayx commented:
@afishbeck please review...

afishbeck commented on tools/esdlcmd-xml/esdl2xml.cpp:
```diff
@@ -33,6 +35,7 @@ void static usage(const char* programName)
 
     puts("Available options:");
     puts(" -r|--recursive: process all includes");
+    puts(" -i|--includePath <include path>: Locations to look for included esdl files");
```
eclcc and eclcmd use -I maybe we should be consistent?

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 81dd7cd910c080e81c222ea36c932a564f98aa3d
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.089 seconds |


Install hpccsystems-platform-community_6.4.2-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented on tools/esdlcmd-xml/esdl2xml.cpp:
```diff
@@ -75,13 +83,37 @@ void parseCommandLine(int argc, char* argv[], Esdl2Esxdl * cmd)
         usage(argv[0]);
 }
 
+bool extractEsdlxmlCmdOption(StringBuffer & option, IProperties * globals, const char * envName, const char * propertyName, const char * defaultPrefix, const char * defaultSuffix)
```
is this the same as "extractEsdlCmdOption" below?  can they be commoned up?

afishbeck commented on tools/esdlcmd/esdlcmd_common.hpp:
```diff
@@ -103,7 +103,11 @@ typedef IEsdlCommand *(*EsdlCommandFactory)(const char *cmdname);
 #define ESDL_OPTION_ECL_HEADER_BLOCK    "--ecl-header"
 
-
+#define ESDLOPT_INCLUDE_PATH            "--includePath"
```
in the name of consistency, eclcmd and esdlcmd normally use the format "--include-path".  Eclcc is a bit inconsistent.

afishbeck commented on tools/esdlcomp/esdlcomp.cpp:
```diff
@@ -1109,22 +1109,20 @@ ESDLcompiler::ESDLcompiler(const char * sourceFile, bool generatefile, const cha
     StringBuffer prot;
     splitFilename(sourceFile, &prot, &srcDir, &name, &ext);
 
+    if (includePath_ && *includePath_)
```
Should you just go ahead and parse this into an array here?

afishbeck replied:
@mayx any comment on the idea of parsing this into a string array right away?

afishbeck commented on tools/esdlcomp/esdlcomp.cpp:
```diff
@@ -1154,6 +1152,61 @@ ESDLcompiler::ESDLcompiler(const char * sourceFile, bool generatefile, const cha
+                    filepath.set(pathInIncludeFull.str());
+                    return true;
+                }
+                VStringBuffer altPathInIncludeFull("%s%s", pathInInclude.str(), alt_ext);
```
You could reuse the existing variable and do 

pathInIncludeFull.setf("%s%s", pathInInclude.str(), alt_ext);

here.

afishbeck commented on tools/esdlcomp/esdlcomp.cpp:
```diff
@@ -1109,22 +1109,20 @@ ESDLcompiler::ESDLcompiler(const char * sourceFile, bool generatefile, const cha
-                alternateExtFilename.append(ESDL_FILE_EXTENSION);
-
-            yyin = fopen(alternateExtFilename.str(), "rt");
+            StringBuffer includedFilePath;
```
there's only a little 'd' between the names includeFilePath and includedFilePath.  Maybe this variable could be called "locatedFilePath"?

afishbeck reviewed (changes requested):
@mayx a few comments.

afishbeck commented:
@mayx plus the multiple -I support we discussed offline.

mayx commented:
@afishbeck those are good suggestions, will make changes accordingly...

mayx commented:
@afishbeck changes made, please review again...

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 82bd4936a5ace959bac2c72e9aca9c56f783de9c
Build: failed
Number of Error(s): 6
collect2: error: ld returned 1 exit status
make[2]: *** [Release/bin/esdl-xml] Error 1
make[1]: *** [tools/esdlcmd-xml/CMakeFiles/esdl-xml.dir/all] Error 2
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
make: *** [all] Error 2
Get ECLWatch build result

afishbeck reviewed (commented):
.

afishbeck reviewed (approved):
@mayx looks good.  Not sure about the build errors?

mayx committed:
HPCC-18132 - ESDL support include paths for searching EspInclude files

- Add the -i --includePath options to relevant esdl commands
- Add support in esdl compiler to search included files in the include
  paths

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 631590fe2a04b66807fbb97f169471492e2a0159
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.294 seconds |


Install hpccsystems-platform-community_6.4.2-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck reviewed (approved):
@mayx looks good
